content/tokio/topics/tracing.md

```diff
@@ -0,0 +1,116 @@
+
+In asynchronous systems like Tokio, interpreting traditional log messages can 
+often be quite challenging. Since individual tasks are multiplexed on the same 
+thread, associated events and log lines are intermixed making it difficult to 
+trace the logic flow. `tracing` expands upon logging-style diagnostics by 
+allowing libraries and applications to record structured events with additional 
+information about *temporality* and *causality* — unlike a log message, a span 
+in `tracing` has a beginning and end time, may be entered and exited by the 
+flow of execution, and may exist within a nested tree of similar spans. In 
+addition, `tracing` spans are *structured*, with the ability to record typed 
+data as well as textual messages.
+
+You can use `tracing` to:
+- observe your application with Open Telemetry-supporting services
+- debug your application with Tokio Console
+- emit logs to `stdout` or `journald`
+- profile where your application is spending time
+
+[`tracing`]: https://docs.rs/tracing
+[`tracing-subscriber`]: https://docs.rs/tracing-subscriber
+
+# Setup
+
+To begin, add [`tracing`] and [`tracing-subscriber`] as dependencies:
+
+```toml
+[dependencies]
+tracing = "0.1.34"
+tracing-subscriber = "0.3"
+```
+
+The [`tracing`] crate provides the API we will use to emit traces. The 
+[`tracing-subscriber`] crate provides some basic utilities for forwarding those 
+traces to external listeners (e.g., `stdout`).
+
+# Subscribing to Traces
+
+If you are authoring an executable (as opposed to a library), you will need to 
+register a tracing *subscriber*. Subscribers are types that process traces 
+emitted by your application and its dependencies, and and can perform tasks 
+such as computing metrics, monitoring for errors, and re-emitting traces to the 
+outside world (e.g., `journald`, `stdout`, or an `open-telemetry` daemon).
+
+In most circumstances, you should register your tracing subscriber as early as
+possible in your `main` function. For instance, the [`FmtSubscriber`] type
+provided by [`tracing-subscriber`] prints formatted traces and events to
+`stdout`, can be registered like so:
+
+```rust
+#[tokio::main]
+pub async fn main() -> mini_redis::Result<()> {
+    // construct a subscriber that prints formatted traces to stdout
+    let subscriber = tracing_subscriber::FmtSubscriber::new();
+    // use that subscriber to process traces emitted after this point
+    tracing::subscriber::set_global_default(subscriber)?;
+
+    ...
+}
+```
+
+[`FmtSubscriber`]: https://docs.rs/tracing-subscriber/latest/tracing_subscriber/fmt/index.html
+
+If you run your application now, you may see some trace events emitted by tokio,
+but you will need to modify your own application to emit traces to get the most
+out of `tracing`.
+
+# Emitting Traces
+
+The easiest way to emit traces is with the [`instrument`] proc-macro annotation
+provided by [`tracing`], which re-writes the bodies of functions to emit traces
+each time they are invoked.
```

Comment:
i would probably _start_ by showing events, and then move on to also showing spans. it might even be worthwhile to show some sample output from the events and spans we add, so that the user knows what to expect --- for example, the default `FmtSubscriber` configuration doesn't print a log line when a span is entered/exited --- it only displays the current spans as context when an event is recorded. people have sometimes been surprised by this, so it would be nice if the tutorial explained that up front.

Reply:
The other bit—it doesn't even technically emit _trace_ each time they're invoked—depending on the function being instrumented, it can either be a _new_ span or a _re-entered_ span, if it's an async function. While I see the utility of this guide, I think that a guide that provides slightly incorrect/imprecise documentation is a concern.

Reply:
mini-redis _does_ have events in it, though, and even briefly discusses the event macro syntax: https://github.com/tokio-rs/mini-redis/blob/ebe4e1f33113d8e43a68199d04943b2581b33604/src/server.rs#L358-L367

we don't need to go into great detail on the syntax and different ways to use the macros; we could just refer to the reference-level documentation here: https://docs.rs/tracing/latest/tracing/#using-the-macros

however, i think this documentation is incomplete if we don't at least _mention_ the following:
- tracing has something called an "event" in it
- events represent moments in time while spans have starts and ends,
- events are recorded using function-like macros (link to docs, one or two examples)

it doesn't have to be a _lot_ of discussion of events, but we should at least say that they exist...